processing/format_qaqc_driver.py

```diff
@@ -0,0 +1,232 @@
+                        pool.close()
+                        pool.join()
+                        process_id_mapping.setdefault(upload_id, process_id)
+                    else:
+                        for upload_id, v in tasks.items():
+                            task, _ = v.values()
+                            log.write((f"Start run: log id {task.upload_id}, "
+                                       f"prior id: {task.prior_process_id}, "
+                                       f"zip id: {task.zip_process_id}, "
+                                       f"run type: {task.run_type}\n"))
+                            process_id, is_upload_sucessful, sub_uuid = \
+                                self.upload_checks.run(task.filename,
+                                                       task.upload_id,
+                                                       task.run_type,
+                                                       task.site_id,
+                                                       task.prior_process_id,
+                                                       task.zip_process_id)
+                            process_id_mapping.setdefault(upload_id,
+                                                          process_id)
+                            sub_uuid_mapping.setdefault(upload_id, None)
+                            if sub_uuid:
+                                sub_uuid_mapping[upload_id] = sub_uuid
+
+                    for upload_id, process_id in process_id_mapping.items():
+                        is_success = \
+                            self.db.check_status_of_process_id(self.conn,
+                                                               process_id)
+                        if is_success:
+                            del tasks[upload_id]
+                            log.write((f"Log id: {upload_id} "
+                                       "ended sucessfully\n"))
+                        else:
+                            retry = tasks[upload_id]['retry']
+                            if retry < self.max_retries:
+                                tasks[upload_id]['retry'] += 1
+                                log.write((f"Log id: {upload_id}, "
+                                           f"retry for {retry} "
+                                           "times and failed\n"))
+                            else:
+                                del tasks[upload_id]
+                                log.write((f"Log id: {upload_id} "
+                                           "ended unsucessfully\n"))
+                    (tasks,
+                     mail_uuid) = self.get_new_data_upload_log(log)
+                    mail_uuids.update(mail_uuid)
+
+                # here check all tasks completed
+                if not self.is_test:
+                    while True:
+                        if self.db.is_all_task_done(self.conn):
+                            break
```

Comment:
My understanding is that we are waiting for all the tasks to be completed.
Can we find a way to be monitoring each upload_token's tasks separately?